Sources/MapboxNavigation/SpriteRepository.swift

```diff
@@ -0,0 +1,147 @@
+import UIKit
+import MapboxMaps
+import MapboxCoreNavigation
+import MapboxDirections
+
+class SpriteRepository {
+    let imageCache: BimodalImageCache
+    let metadataCache =  SpriteMetaDataCache()
+    var styleURI: StyleURI = .navigationDay
+    var baseURL: URL = URL(string: "https://api.mapbox.com/styles/v1")!
+    
+    public var sessionConfiguration: URLSessionConfiguration = URLSessionConfiguration.default {
+        didSet {
+            imageDownloader = ImageDownloader(sessionConfiguration: sessionConfiguration)
+        }
+    }
+
+    public static let shared = SpriteRepository.init()
+    fileprivate(set) var imageDownloader: ReentrantImageDownloader
+    
+    init(imageCache: BimodalImageCache = ImageCache(), withDownloader downloader: ReentrantImageDownloader = ImageDownloader()) {
+        self.imageCache = imageCache
+        self.imageDownloader = downloader
+    }
+    
+    func updateRepository(styleURI: StyleURI? = nil, shield: VisualInstruction.Component.ShieldRepresentation? = nil, completion: @escaping CompletionHandler) {
+        let baseURL = shield?.baseURL ?? self.baseURL
+        let styleURI = styleURI ?? self.styleURI
+        
+        guard let styleID = styleURI.rawValue.components(separatedBy: "styles")[safe: 1],
+              let spriteRequestURL = spriteURL(isImage: true, baseURL: baseURL, styleID: styleID),
+              let metadataRequestURL = spriteURL(isImage: false, baseURL: baseURL, styleID: styleID) else {
+                  return
+              }
+
+        let dispatchGroup = DispatchGroup()
+        
+        dispatchGroup.enter()
+        downloadMetadata(metadataRequestURL) { (_) in
+            dispatchGroup.leave()
+        }
+        
+        dispatchGroup.enter()
+        downLoadSprite(spriteRequestURL) { (_) in
+            dispatchGroup.leave()
+        }
+        
+        dispatchGroup.notify(queue: .main) {
+            self.styleURI = styleURI
+            self.baseURL = baseURL
+            completion()
+        }
+    }
+    
+    func spriteURL(isImage: Bool, baseURL: URL, styleID: String) -> URL? {
+        guard var urlComponent = URLComponents(url: baseURL, resolvingAgainstBaseURL: false),
+              let accessToken = NavigationSettings.shared.directions.credentials.accessToken else { return nil }
+        
+        let requestTyep = isImage ? "/sprite@2x.png" : "/sprite@2x"
+        urlComponent.path += styleID
+        urlComponent.path += requestTyep
+        urlComponent.queryItems = [URLQueryItem(name: "access_token", value: accessToken)]
+        return urlComponent.url
+    }
+    
+    func downloadMetadata(_ metadataURL: URL, completion: @escaping (Data?) -> Void) {
```

Comment:
SpriteRepository is functioning similarly to the Directions class of MapboxDirections or the [Snapshot](https://github.com/mapbox/MapboxStatic.swift/blob/083e3daaa60ecdc142ca2e0360288cd233d4c851/Sources/MapboxStatic/Snapshot.swift#L192) class of MapboxStatic, so as tail work, it would be great to align the method signatures for consistency and future extensibility. For example, this method could return a URLSessionDataTask to enable canceling redundant requests.

Reply:
Thanks for the suggestion, I'll make it a tail work.